File: test/gitlab.test.ts
import { describe, expect, it, vi } from 'vitest';
import {
  createMergeRequest,
  postMrNote,
  waitForMrPipeline,
  type GitlabCredentials,
  type GitlabHttpClient,
} from '../src/integrations/gitlab.js';

const creds: GitlabCredentials = { baseUrl: 'https://gitlab.com', projectId: 42, token: 'glpat-x' };

/** Fake http client recording calls and replaying scripted responses. */
function fakeClient(responses: Array<{ status: number; body?: unknown }>) {
  const calls: Array<{ url: string; init: { method: string; headers: Record<string, string>; body?: string } }> = [];
  const http: GitlabHttpClient = {
    request: async (url, init) => {
      calls.push({ url, init });
      const next = responses.shift();
      if (!next) throw new Error('unexpected extra HTTP call');
      return {
        ok: next.status >= 200 && next.status < 300,
        status: next.status,
        json: async () => next.body,
        text: async () => (next.body === undefined ? '' : JSON.stringify(next.body)),
      };
    },
  };
  return { calls, http };
}

describe('createMergeRequest', () => {
  it('posts PRIVATE-TOKEN MR and returns web_url', async () => {
    const { http, calls } = fakeClient([
      { status: 201, body: { web_url: 'https://gitlab.com/g/p/-/merge_requests/3', iid: 3 } },
    ]);
    const url = await createMergeRequest(
      creds,
      { sourceBranch: 'devagent/E-1', targetBranch: 'main', title: 't', description: 'd' },
      http,
    );
    expect(url).toBe('https://gitlab.com/g/p/-/merge_requests/3');
    expect(calls[0]?.url).toContain('/api/v4/projects/42/merge_requests');
    expect(calls[0]?.init.headers['PRIVATE-TOKEN']).toBe('glpat-x');
    const body = JSON.parse(String(calls[0]?.init.body)) as Record<string, string>;
    expect(body.source_branch).toBe('devagent/E-1');
    expect(body.target_branch).toBe('main');
  });

  it('URL-encodes project path ids', async () => {
    const { http, calls } = fakeClient([{ status: 201, body: { web_url: 'u' } }]);
    await createMergeRequest(
      { ...creds, projectId: 'group/proj' },
      { sourceBranch: 'a', targetBranch: 'b', title: 't', description: 'd' },
      http,
    );
    expect(calls[0]?.url).toContain(encodeURIComponent('group/proj'));
  });

  it('throws with HTTP status and API error body on failure', async () => {
    const { http } = fakeClient([{ status: 401, body: { message: 'invalid token' } }]);
    await expect(
      createMergeRequest(creds, { sourceBranch: 'a', targetBranch: 'b', title: 't', description: 'd' }, http),
    ).rejects.toThrow(/HTTP 401: invalid token/);
  });

  it('throws when web_url is missing', async () => {
    const { http } = fakeClient([{ status: 201, body: {} }]);
    await expect(
      createMergeRequest(creds, { sourceBranch: 'a', targetBranch: 'b', title: 't', description: 'd' }, http),
    ).rejects.toThrow(/web_url/);
  });
});

describe('postMrNote', () => {
  it('posts note body to the MR notes endpoint', async () => {
    const { http, calls } = fakeClient([{ status: 201, body: { id: 9 } }]);
    await postMrNote(creds, 3, 'validation evidence: all gates green', http);
    expect(calls[0]?.url).toBe('https://gitlab.com/api/v4/projects/42/merge_requests/3/notes');
    expect(JSON.parse(String(calls[0]?.init.body)).body).toContain('validation evidence');
  });

  it('throws with API error detail on failure', async () => {
    const { http } = fakeClient([{ status: 404, body: { error: '404 Not Found' } }]);
    await expect(postMrNote(creds, 3, 'x', http)).rejects.toThrow(/HTTP 404: 404 Not Found/);
  });
});

describe('waitForMrPipeline', () => {
  const noSleep = async () => {};

  it('resolves true when the latest pipeline succeeds', async () => {
    const { http } = fakeClient([{ status: 200, body: [{ id: 5, status: 'success' }] }]);
    await expect(waitForMrPipeline(creds, 3, http, { sleep: noSleep })).resolves.toBe(true);
  });

  it('polls until terminal state, sleeping between attempts', async () => {
    const sleeps: number[] = [];
    const { http } = fakeClient([
      { status: 200, body: [{ id: 1, status: 'running' }] },
      { status: 200, body: [{ id: 1, status: 'pending' }] },
      { status: 200, body: [{ id: 1, status: 'success' }] },
    ]);
    await expect(waitForMrPipeline(creds, 3, http, { pollIntervalMs: 2_000, sleep: async (ms) => { sleeps.push(ms); } })).resolves.toBe(true);
    expect(sleeps).toEqual([2_000, 2_000]);
  });

  it('throws listing pipeline statuses on terminal failure', async () => {
    const { http } = fakeClient([{ status: 200, body: [{ id: 6, status: 'failed' }, { id: 5, status: 'success' }] }]);
    await expect(waitForMrPipeline(creds, 3, http, { sleep: noSleep })).rejects.toThrow(/#6:failed/);
  });

  it('times out loudly while pipelines stay non-terminal', async () => {
    const { http } = fakeClient(Array.from({ length: 30 }, () => ({ status: 200 as const, body: [{ id: 1, status: 'running' }] })));
    await expect(waitForMrPipeline(creds, 3, http, { timeoutMs: 0, pollIntervalMs: 1, sleep: noSleep })).rejects.toThrow(/timed out/);
  });

  it('waits (not fails) when no pipelines exist yet', async () => {
    const { http } = fakeClient([
      { status: 200, body: [] },
      { status: 200, body: [{ id: 2, status: 'success' }] },
    ]);
    await expect(waitForMrPipeline(creds, 3, http, { pollIntervalMs: 1, sleep: noSleep })).resolves.toBe(true);
  });
});

File: internal/integrations/gitlab_test.go
// Contract tests for the GitLab publisher, ported from test/gitlab.test.ts
// with recorded fixtures. All HTTP is seamed through Doer — zero network.
package integrations

import (
	"context"
	"reflect"
	"strings"
	"testing"
)

var glCreds = GitlabCredentials{BaseURL: "https://gitlab.com", ProjectID: "42", Token: "glpat-x"}

func TestGitlabCreateMergeRequestPostsPrivateToken(t *testing.T) {
	// Ported: posts PRIVATE-TOKEN MR and returns web_url.
	tr := newRecordedTransport(t, recordedResponse(201, nil, fixtureString(t, "gitlab-mr-create-response.json")))
	url, err := CreateMergeRequest(context.Background(), glCreds,
		CreateMrOptions{SourceBranch: "devagent/E-1", TargetBranch: "main", Title: "t", Description: "d"}, tr)
	if err != nil {
		t.Fatalf("CreateMergeRequest: %v", err)
	}
	if url != "https://gitlab.com/g/p/-/merge_requests/3" {
		t.Errorf("url = %q", url)
	}
	req := tr.calls[0]
	if !strings.Contains(req.URL.String(), "/api/v4/projects/42/merge_requests") {
		t.Errorf("url = %q", req.URL)
	}
	// The PRIVATE-TOKEN key is stored verbatim so the wire carries it
	// un-canonicalized, byte-equal to the TS request.
	if got := req.Header["PRIVATE-TOKEN"]; len(got) != 1 || got[0] != "glpat-x" {
		t.Errorf("PRIVATE-TOKEN = %#v", req.Header)
	}
	// Publisher payload byte-parity against the recorded request fixture.
	if string(tr.bodies[0]) != fixtureString(t, "gitlab-mr-create-request.json") {
		t.Errorf("MR payload mismatch:\n got  %s\n want %s", tr.bodies[0], fixtureString(t, "gitlab-mr-create-request.json"))
	}
	parsed := graphqlBody(t, tr.bodies[0])
	if parsed["source_branch"] != "devagent/E-1" {
		t.Errorf("source_branch = %v", parsed["source_branch"])
	}
	if parsed["target_branch"] != "main" {
		t.Errorf("target_branch = %v", parsed["target_branch"])
	}
}

func TestGitlabCreateMergeRequestURLEncodesProjectPath(t *testing.T) {
	// Ported: URL-encodes project path ids.
	tr := newRecordedTransport(t, recordedResponse(201, nil, `{"web_url":"u"}`))
	_, err := CreateMergeRequest(context.Background(),
		GitlabCredentials{BaseURL: "https://gitlab.com", ProjectID: "group/proj", Token: "glpat-x"},
		CreateMrOptions{SourceBranch: "a", TargetBranch: "b", Title: "t", Description: "d"}, tr)
	if err != nil {
		t.Fatal(err)
	}
	// encodeURIComponent("group/proj") = "group%2Fproj".
	if !strings.Contains(tr.calls[0].URL.String(), "group%2Fproj") {
		t.Errorf("url = %q", tr.calls[0].URL)
	}
}

func TestGitlabCreateMergeRequestThrowsWithAPIErrorBody(t *testing.T) {
	// Ported: throws with HTTP status and API error body on failure.
	tr := newRecordedTransport(t, recordedResponse(401, nil, fixtureString(t, "gitlab-mr-error.json")))
	_, err := CreateMergeRequest(context.Background(), glCreds,
		CreateMrOptions{SourceBranch: "a", TargetBranch: "b", Title: "t", Description: "d"}, tr)
	if err == nil || !strings.Contains(err.Error(), "HTTP 401: invalid token") {
		t.Errorf("err = %v", err)
	}
}

func TestGitlabCreateMergeRequestThrowsWhenWebURLMissing(t *testing.T) {
	// Ported: throws when web_url is missing.
	tr := newRecordedTransport(t, recordedResponse(201, nil, fixtureString(t, "gitlab-mr-missing-web-url.json")))
	_, err := CreateMergeRequest(context.Background(), glCreds,
		CreateMrOptions{SourceBranch: "a", TargetBranch: "b", Title: "t", Description: "d"}, tr)
	if err == nil || !strings.Contains(err.Error(), "web_url") {
		t.Errorf("err = %v", err)
	}
}

func TestGitlabPostMrNote(t *testing.T) {
	// Ported: posts note body to the MR notes endpoint.
	tr := newRecordedTransport(t, recordedResponse(201, nil, `{"id":9}`))
	if err := PostMrNote(context.Background(), glCreds, 3, "validation evidence: all gates green", tr); err != nil {
		t.Fatalf("PostMrNote: %v", err)
	}
	if got := tr.calls[0].URL.String(); got != "https://gitlab.com/api/v4/projects/42/merge_requests/3/notes" {
		t.Errorf("url = %q", got)
	}
	// Note payload byte-parity against the recorded request fixture.
	if string(tr.bodies[0]) != fixtureString(t, "gitlab-mr-note-request.json") {
		t.Errorf("note payload mismatch:\n got  %s\n want %s", tr.bodies[0], fixtureString(t, "gitlab-mr-note-request.json"))
	}
}

func TestGitlabPostMrNoteThrowsWithAPIErrorDetail(t *testing.T) {
	// Ported: throws with API error detail on failure.
	tr := newRecordedTransport(t, recordedResponse(404, nil, fixtureString(t, "gitlab-mr-note-error.json")))
	err := PostMrNote(context.Background(), glCreds, 3, "x", tr)
	if err == nil || !strings.Contains(err.Error(), "HTTP 404: 404 Not Found") {
		t.Errorf("err = %v", err)
	}
}

func TestGitlabWaitForMrPipelineResolvesOnSuccess(t *testing.T) {
	// Ported: resolves true when the latest pipeline succeeds.
	tr := newRecordedTransport(t, recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-success.json")))
	ok, err := WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{PollIntervalMs: -1, Sleep: noopSleep, Doer: tr})
	if err != nil || !ok {
		t.Errorf("ok = %v, err = %v", ok, err)
	}
}

func TestGitlabWaitForMrPipelinePollsUntilTerminal(t *testing.T) {
	// Ported: polls until terminal state, sleeping between attempts.
	tr := newRecordedTransport(t,
		recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-running.json")),
		recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-pending.json")),
		recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-success.json")),
	)
	var sleeps []int
	ok, err := WaitForMrPipeline(context.Background(), glCreds, 3,
		PipelineWaitOptions{PollIntervalMs: 2000, Sleep: func(ms int) { sleeps = append(sleeps, ms) }, Doer: tr})
	if err != nil || !ok {
		t.Fatalf("ok = %v, err = %v", ok, err)
	}
	if !reflect.DeepEqual(sleeps, []int{2000, 2000}) {
		t.Errorf("sleeps = %#v", sleeps)
	}
}

func TestGitlabWaitForMrPipelineThrowsListingStatuses(t *testing.T) {
	// Ported: throws listing pipeline statuses on terminal failure.
	tr := newRecordedTransport(t, recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-failed.json")))
	_, err := WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{PollIntervalMs: -1, Sleep: noopSleep, Doer: tr})
	if err == nil || !strings.Contains(err.Error(), "#6:failed") {
		t.Errorf("err = %v", err)
	}
}

func TestGitlabWaitForMrPipelineTimesOutLoudly(t *testing.T) {
	// Ported: times out loudly while pipelines stay non-terminal.
	// TimeoutMs: -1 puts the deadline in the past so the first
	// non-terminal poll times out (the vitest case used timeoutMs 0).
	tr := newRecordedTransport(t, recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-running.json")))
	_, err := WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{TimeoutMs: -1, PollIntervalMs: -1, Sleep: noopSleep, Doer: tr})
	if err == nil || !strings.Contains(err.Error(), "timed out") {
		t.Errorf("err = %v", err)
	}
}

func TestGitlabWaitForMrPipelineWaitsWhenNoPipelinesYet(t *testing.T) {
	// Ported: waits (not fails) when no pipelines exist yet.
	tr := newRecordedTransport(t,
		recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-empty.json")),
		recordedResponse(200, nil, fixtureString(t, "gitlab-pipelines-success.json")),
	)
	ok, err := WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{PollIntervalMs: -1, Sleep: noopSleep, Doer: tr})
	if err != nil || !ok {
		t.Errorf("ok = %v, err = %v", ok, err)
	}
}

func TestGitlabWaitForMrPipelineNonArrayBodyDegradesToEmpty(t *testing.T) {
	// Mirrors `Array.isArray(pipelines) ? pipelines : []`: a valid-JSON
	// non-array body degrades to an empty list (then the deadline check
	// times out), while invalid JSON is a decode error.
	tr := newRecordedTransport(t, recordedResponse(200, nil, `{"unexpected":"shape"}`))
	_, err := WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{TimeoutMs: -1, PollIntervalMs: -1, Sleep: noopSleep, Doer: tr})
	if err == nil || !strings.Contains(err.Error(), "timed out") {
		t.Errorf("err = %v", err)
	}
	tr2 := newRecordedTransport(t, recordedResponse(200, nil, `not-json`))
	_, err = WaitForMrPipeline(context.Background(), glCreds, 3, PipelineWaitOptions{PollIntervalMs: -1, Sleep: noopSleep, Doer: tr2})
	if err == nil || !strings.Contains(err.Error(), "invalid GitLab pipeline response") {
		t.Errorf("err = %v", err)
	}
}
